### PR TITLE
Enabled wells with poromechanics

### PR DESCRIPTION
**Pull request by francoishamon**
This PR enables the coupling between a poromechanics solver and a well solver, such as:

- [X] `MultiphasePoromechanics` + `CompositionalMultiphaseWell`
- [X] `SinglePhasePoromechanics` + `SinglePhaseWell`

This is done mostly by uncommenting the catalog names that had been prepared in #1890. 

The PR also adds two integrated tests (single-phase and multiphase) coupled poromechanics and well. 
The MGR recipes will be added in a separate PR.  

Rebaseline done in https://github.com/GEOSX/integratedTests/pull/256

### CODE REVIEW

**Review by TotoGaz — COMMENTED** [review]
May be step-by-step settings here and there.

File: src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp [hunk 1]
```diff
@@ -82,29 +82,24 @@ CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
...
+  if( auto const * const ptr = dynamicCast< MultiphasePoromechanicsSolver const * >( this->reservoirSolver() ) )
   {
+    useMassFlow = ptr->flowSolver()->template getReference< integer >( CompositionalMultiphaseBase::viewKeyStruct::useMassFlagString() );
+    Base::wellSolver()->setFlowSolverName( ptr->flowSolver()->getName() );
+    // TODO: add an MGR recipe here
+  }
+  else // the reservoir solver is a flow solver
+  {
+    useMassFlow = Base::reservoirSolver()->template getReference< integer >( CompositionalMultiphaseBase::viewKeyStruct::useMassFlagString() );
+    Base::wellSolver()->setFlowSolverName( Base::m_names[toUnderlying( Base::SolverType::Reservoir )] );
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp:102]: I done multiple times, yes, probably. I let @francoishamon figures this for us!

File: src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp [hunk 2]
```diff
@@ -133,6 +133,7 @@ void CompositionalMultiphaseWell::registerDataOnMesh( Group & meshBodies )
     m_numComponents = fluid0.numFluidComponents();
   }
   m_numDofPerWellElement = m_numComponents + 2; // 1 pressure + NC compositions + 1 connectionRate
+  m_numDofPerResElement = m_numComponents + 1;
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp]: I added the comment for the new line as well

File: src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp [hunk 3]
```diff
@@ -82,29 +82,24 @@ CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
...
-                  InputError );
-}
-
 template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
 void
 CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
 initializePreSubGroups()
 {
-  if( catalogName() == CompositionalCatalogNames< CompositionalMultiphaseBase >::name() )
+  integer useMassFlow = 0;
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp]: In the latest commit `useMassFlow` is now a `bool`

File: src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp [hunk 4]
```diff
@@ -47,16 +48,15 @@ template< typename COMPOSITIONAL_RESERVOIR_SOLVER > class
 template<> class CompositionalCatalogNames< CompositionalMultiphaseBase >
 {
 public:
+  // TODO: find a way to use the catalog name here
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp:51]: I mean that it is be done by assembling the coupled solver catalog name using the single-physics catalog names, as described here (https://github.com/GEOSX/GEOSX/pull/1890#discussion_r909064781) but this is going to change many xml files, so I prefer doing it when we address the renaming issue https://github.com/GEOSX/GEOSX/issues/2003

File: src/coreComponents/physicsSolvers/multiphysics/SinglePhaseReservoirAndWells.cpp [hunk 5]
```diff
@@ -82,8 +82,16 @@ void
...
+
+  // if the reservoir solver is a coupled solver itself
+  if( auto const * const ptr = dynamicCast< SinglePhasePoromechanicsSolver const * >( this->reservoirSolver() ) )
+  {
+    Base::wellSolver()->setFlowSolverName( ptr->flowSolver()->getName() );
+    // TODO: add an MGR recipe here
+  }
+  else // the reservoir solver is a flow solver
   {
+    Base::wellSolver()->setFlowSolverName( Base::m_names[toUnderlying( Base::SolverType::Reservoir )] );
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/multiphysics/SinglePhaseReservoirAndWells.cpp]: Done, I changed this function quite a bit, now the flow solver name is defined in one pass like this:
https://github.com/GEOSX/GEOSX/blob/f83e8c1eacc389b317f20bce1959be8920cea585/src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp#L131
where `flowSolver` is a function that returns a pointer of type `CompositionalMultiphaseBase const*`

File: src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp]: Done, now this is just:
https://github.com/GEOSX/GEOSX/blob/f83e8c1eacc389b317f20bce1959be8920cea585/src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp#L130-L132
where `flowSolver` is a function that returns a pointer of type `CompositionalMultiphaseBase const*` and `setMGRStrategy` is based on the single-physics MGR strategy type. In other words, I removed the `dynamicCast` entirely.

**Review by TotoGaz — APPROVED** [review]

**Pull request merged**